### PR TITLE
Feature: Open with Jetbrains IDEs (Autodetected)

### PR DESCRIPTION
**Pull request by ennerperez**
image

Automatic JetBrains IDE detection using Toolbox installed tools or using environment variables:

JETBRAINS_RD_PATH => Rider
JETBRAINS_WS_PATH => WebStorm
JETBRAINS_PS_PATH => PhpStorm
JETBRAINS_DB_PATH => DataGrip
JETBRAINS_IU_PATH => IntelliJ IDEA Ultimate
JETBRAINS_WRS_PATH => Writerside
JETBRAINS_IC_PATH => IntelliJ IDEA Community Edition
JETBRAINS_PC_PATH => PyCharm Community
JETBRAINS_RM_PATH => RubyMine
JETBRAINS_QA_PATH => Aqua
JETBRAINS_DS_PATH => DataSpell
JETBRAINS_PY_PATH => PyCharm Professional
JETBRAINS_GO_PATH => GoLand
JETBRAINS_RR_PATH => RustRover

### CODE REVIEW

**Comment by love-linger** [comment]
I've merge this PR into `develop` branch, and modified some of your code. 

## code_review:

* use JsonSerializerContext to avoid AOT warnnings
* since we call TryAdd by interating the installed tools, so detecting by environment variable becomes meaningless (it can not detect tools not installed by Toolbox). Just add it into founded directly
* remove unnecessary type defines
* determine the Icon used by tool while adding it to the founded list.

## resolved conflicts:
* src/Native/Linux.cs
* src/Native/MacOS.cs
* src/Native/Windows.cs

**Comment by love-linger** [comment]
One more thing:

* How to detect tools not installed by the JetBrains Toolbox? In your original code, TryAdd is called by iterating the tools installed by JetBrains Toolbox, so environment variable becomes meaningless. I've remove it, but the issue remains.

**Comment by love-linger** [comment]
I need to close this PR first. Please open new PR, and make sure the target branch is `develop` instead of `master`

**Comment by love-linger** [comment]
I've found that if user installed JetBrains Toolbox, it will automatically detect the JetBrains products and add the results to the state.json. So I add a tip about this to README.md.

File: src/Models/ExternalTool.cs [hunk 1]
```diff
@@ -76,11 +125,70 @@ public void TryAdd(string name, string icon, string args, string env, Func<strin
...
             {
-                Name = name,
-                Icon = icon,
-                OpenCmdArgs = args,
-                Executable = path,
+                Name = name, Icon = icon, OpenCmdArgs = args, Executable = path, FallbackIcon = fallbackIcon
             });
         }
+
+        public void FindJetBrainsFromToolbox(Func<string> platform_finder)
```
Comment (Rwing) [inline src/Models/ExternalTool.cs:132]: “platform_finder” I dont think that's a good naming style :(

Comment (love-linger) [reply src/Models/ExternalTool.cs:132]: It's my mistake.

**Pull request closed without merging**